File: libs/store-dao/src/main/kotlin/mono/store/dao/workspace/WorkspaceDao.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.store.dao.workspace

import mono.store.manager.StorageDocument
import mono.store.manager.StoreKeys.LAST_OPEN
import mono.store.manager.StoreKeys.OBJECT_CONTENT
import mono.store.manager.StoreKeys.PATH_SEPARATOR
import mono.store.manager.StoreKeys.WORKSPACE

/**
 * A dao for workspace
 */
class WorkspaceDao private constructor(
    private val workspaceDocument: StorageDocument
) {
    var lastOpenedObjectId: String?
        get() = workspaceDocument.get(LAST_OPEN)
        set(value) {
            if (value != null) {
                workspaceDocument.set(LAST_OPEN, value)
            }
        }

    private val objectDaos: MutableMap<String, WorkspaceObjectDao> = mutableMapOf()

    fun getObject(objectId: String): WorkspaceObjectDao = objectDaos.getOrPut(objectId) {
        WorkspaceObjectDao(objectId, workspaceDocument)
    }

    fun removeObject(objectId: String) {
        getObject(objectId).removeSelf()
        // TODO: 
        //  If the object is currently open, choose a latest opened object and make it active.
        //  If no object left, create a blank workspace.
        objectDaos.remove(objectId)
    }

    /**
     * Gets list of all objects in the storage, ordered by last opened time desc.
     */
    fun getObjects(): Sequence<WorkspaceObjectDao> =
        workspaceDocument.getKeys {
            it.startsWith(WORKSPACE + PATH_SEPARATOR) &&
                it.endsWith(PATH_SEPARATOR + OBJECT_CONTENT)
        }
            .map { getObject(objectId = it.split(PATH_SEPARATOR)[1]) }
            .sortedByDescending { it.lastOpened }

    companion object {
        val instance: WorkspaceDao by lazy { WorkspaceDao(StorageDocument.get(WORKSPACE)) }
    }
}

File: monosketch-svelte/src/lib/mono/store-manager/dao/workspace-dao.ts
import { Flow } from "$libs/flow";
import { StorageDocument, StoreKeys } from "$mono/store-manager";
import { WorkspaceObjectDao } from "$mono/store-manager/dao/workspace-object-dao";

/**
 * A dao for workspace
 */
export class WorkspaceDao {
    private static _instance: WorkspaceDao | null = null;

    private readonly workspaceDocument: StorageDocument;
    private objectDaos: Map<string, WorkspaceObjectDao>;

    private workspaceUpdateMutableFlow: Flow<string> = new Flow('');
    workspaceUpdateFlow: Flow<string> = this.workspaceUpdateMutableFlow.immutable();

    private constructor(workspaceDocument: StorageDocument) {
        this.workspaceDocument = workspaceDocument;
        this.objectDaos = new Map<string, WorkspaceObjectDao>();

        workspaceDocument.setObserver(StoreKeys.LAST_MODIFIED_TIME, () => {
            this.workspaceUpdateMutableFlow.value = this.lastOpenedObjectId ?? '';
        });
    }

    // Accessor property for lastOpenedObjectId
    get lastOpenedObjectId(): string | null {
        return this.workspaceDocument.get(StoreKeys.LAST_MODIFIED_PROJECT_ID);
    }

    set lastOpenedObjectId(value: string | null) {
        if (value !== null) {
            this.workspaceDocument.set(StoreKeys.LAST_MODIFIED_PROJECT_ID, value);
        }
    }

    getObject(objectId: string): WorkspaceObjectDao {
        if (!this.objectDaos.has(objectId)) {
            const objectDao = new WorkspaceObjectDao(objectId, this.workspaceDocument);
            this.objectDaos.set(objectId, objectDao);
        }
        return this.objectDaos.get(objectId)!;
    }

    removeObject(objectId: string): void {
        const objectDao = this.getObject(objectId);
        objectDao.removeSelf();
        // TODO:
        //  If the object is currently open, choose a latest opened object and make it active.
        //  If no object left, create a blank workspace.
        this.objectDaos.delete(objectId);
    }

    // Gets list of all objects in the storage, ordered by last opened time desc.
    getObjects(): WorkspaceObjectDao[] {
        return Array.from(this.workspaceDocument.getKeys(key =>
            key.startsWith(StoreKeys.WORKSPACE + StoreKeys.PATH_SEPARATOR) &&
            key.endsWith(StoreKeys.PATH_SEPARATOR + StoreKeys.OBJECT_CONTENT),
        ))
            .map(key => this.getObject(key.split(StoreKeys.PATH_SEPARATOR)[1]))
            .sort((a, b) => b.lastOpened - a.lastOpened); // Assuming lastOpened is a numeric timestamp
    }

    // Singleton instance accessor
    static get instance(): WorkspaceDao {
        if (!WorkspaceDao._instance) {
            WorkspaceDao._instance = new WorkspaceDao(StorageDocument.get(StoreKeys.WORKSPACE));
        }
        return WorkspaceDao._instance;
    }
}
